docs/design/sql/08-expressions.md

```diff
@@ -0,0 +1,262 @@
+  `BIGINT`); numeric literals containing decimal point (`1.1`, `4.2`,
+  etc.) receive `DECIMAL` type; scientific notation numeric literals
+  (`1e1`, `4.2e2`, etc.) receive `DOUBLE` type; string literals
+  (`'foo'`) receive `VARCHAR` type.
+
+- Parameters: parameter types are inferred from the context: `1.0 + ?`,
+  the parameter would receive `DECIMAL` type since the type of the
+  left-hand side literal is `DECIMAL`; `1 + ?`, despite the fact that
+  the literal type is `TINYINT` the parameter would receive `BIGINT`
+  type because `TINYINT` is too restrictive; `sin(?)`, the parameter
+  would receive `DOUBLE` type since it's the only type accepted by `sin`
+  function. In certain cases (`SELECT ?`, for instance) it's impossible
+  to infer a concrete type, more on that below.
+  
+In general, the validation process begins with a call to
+`SqlValidator.validate`, which starts a recursive validation of the
+passed node and its children. Typically, during the validation, types
+should be _derived_ for every node starting at the root, that usually
+requires knowing types of all child nodes. Types of certain nodes
+(dynamic parameters, for instance) might be unknown, so as a first step
+every node tries to _infer_ types for its child nodes of unknown types.
+For operators this process can be customized using
+`SqlOperandTypeInference` strategy associated with an operator.
+
+If a node type can't be inferred, the validation fails. Currently, that
+happens if an operator acts only on dynamic parameters and/or `NULL`s.
+This might change in the future, for instance we may assign some default
+type in such cases. Consider `? + ?`, `+` operator can be applied to
+numeric and temporal types, we may choose to assign `DECIMAL` type for
+the parameters in this specific case. Alternatively, we might take into
+account the actual types of the passed parameter arguments and construct
+separate query plans based on that information, that way we would not
+need any one-size-fits-all defaults. `NULL` as a type is not a
+first-class citizen in Calcite: basically, every `NULL` literal
+participating in some operator or function must have a concrete type
+assigned, therefore `NULL + NULL` is affected by the same type inference
+problem, which might be solved by making `NULL` type a first-class
+citizen or by choosing some defaults.
+
+After all unknown types are inferred and assigned to nodes, type for
+every tree node can be derived from the node itself potentially
+consulting its child nodes for their types. Most operators have an
+additional round of type refinement called type coercion. For instance,
+both sides of the binary comparison operators are coerced to the same
+type respecting type precedence and conversion rules defined in [Type
+System design document](01-type-system.md). Another example is the
+binary arithmetic operators coercing their operands to a common type.
+The coercion can be customized by providing a custom `TypeCoercion` (see
```

Comment:
AFAIK this is not the main problem. The bigger problem is that we allow implicit conversions for those pairs, that are defined explicitly in the `01-type-system.md`. For example, `LENGTH(intCol)` will just work fine, while INT -> VARCHAR explicit conversion is prohibited. 

I would advise to mention it in the document explicitly:
1. We currently do incorrect implicit conversions that otherwise should be explicit only - these invalid conversions are performed by Calcite code, not ours, and this should be fixed somehow in the future
1. We currently do not understand whether we should have this `implicit/explicit` notion at all, or whether allowed conversions should be defined on the operator level. Or maybe we should have `implicit/explicit` conversions *and* per-operator overrides.

Reply:
fixed in 2311497e69a49d9d38938c7b67de723ef9cbceaf